### PR TITLE
feat: implement @dcl/rpc

### PR DESCRIPTION
**Pull request by leanmendoza**
# What? 
https://github.com/decentraland/sdk/issues/313
Replace the `decentraland-rpc` usage with `@dcl/rpc` in scenes.

# Why? 
`decentraland-rpc` doesn't support binary messages, an important feature to optimize the protocol and to implement the new ECS7.

# How to test
It's possible to test in every environment, just add `&kernel-branch=feat/implement-@dcl-rpc` in the URL. The behavior should be the same as if the query parameter isn't added.

The overhead added (temporary) is:
- JSON.stringify and JSON.parse in the events payload, the most case used is the position report and avatar scenes.
- Each engine scene message was JSON-serialized, now it's JSON-serialized and then protocolbuffer-serialized. In other words, one layer of serialization is **temporary** added, so this can introduce some delays or make it heavier.

Some environment examples:
- Prod https://play.decentraland.org/?position=0,0&kernel-branch=feat/implement-@dcl-rpc
- Zone with ROPSTEN https://play.decentraland.zone/?position=0,0&kernel-branch=feat/implement-@dcl-rpc&NETWORK=ropsten
- SDK test scenes https://sdk-test-scenes.decentraland.io/?kernel-branch=feat/implement-@dcl-rpc
- Your local preview https://127.0.0.1:8000/?kernel-branch=feat/implement-@dcl-rpc

**You have two paths in each environment, as a guest and with a wallet.**

# How the code was changed

## @dcl/rpc host implementation
**Path: `packages/world`**

> In the previous implementation, kernel fed the APIs with `system.getAPIInstance(XXXX)` getting the class implemented and modifying some variable members. 

Now, the `SceneWorker` creates the context with all the data before starting the system. This context is passed in all service methods and it can be patched by accessing the `SceneWorker` object and modifying the `rpcContext` member.

## Scene runtime implementation
**Path: `packages/scene-system`**
The `SceneRuntime.ts` was rewritten, now the scene runtime consumes `event` from the EngineAPI with a pushable channel, instead of native on-emit messages from the worker (decentraland-rpc really).

## APIs 
- Path: `packages/shared/apis`
The changed line counter is significant because of this.

1. All APIs were transcribed to `.proto` as RPC services. 
2. All API calls were modified because the request and response data are only one object instead of multiple parameters, undefined responses or void responses, etc. 
3. To maintain the backward compatibility, most of the API-clients preserved the response and request data structure except EngineAPI (now the events are pushed to the channel, and this channel streamed)
  a. The event payloads now have JSON.stringify on the host-side and JSON.parse on the worker-side, this can introduce some delays  
5. SignedFetch is another modified API, the request object was trimmed to only `method`, `body`, and `headers`.

# Known issues (X indicates that it's solved)
- [x] Delayed event receiving: sometimes it seems some events are enqueued and the delivery delays
- [x] When close the worker, the host throw the error that is expecting the ACK message:
Screenshot_20220606_191024
- [x] Fix avatar tests
- [x] Fix circular dependency (With StatefulWorker)

### CODE REVIEW

File: packages/shared/apis/SceneStateStorageController.proto [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+message SaveProjectInfoResponse {
+
+}
+
+message PublishSceneStateRequest {
+
+}
+
+message PublishSceneStateResponse {
```
Comment (menduz) [inline packages/shared/apis/SceneStateStorageController.proto:92]: are these empty messages ok?

Comment (leanmendoza) [reply packages/shared/apis/SceneStateStorageController.proto:96]: done

File: packages/scene-system/stateful.scene.system.ts [hunk 2]
```diff
@@ -1,136 +1,143 @@
...
+//       this.devToolsAdapter.log(...messages)
+//     } else {
+//       defaultLogger.info('', ...messages)
+//     }
+//   }
+// }
+
+// new StatefulWebWorkerScene(WebWorkerTransport(self))
+
+throw new Error('Deprecated New RPC, stateful.scene.system.ts needs to be reimplemented with new RPC.')
```
[the remark this answers is not in the document]
Comment (leanmendoza) [reply packages/scene-system/stateful.scene.system.ts]: it's just there now :smile:

File: packages/shared/apis/client/index.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// import './SocialController'
+// import './DevTools'
+// import './CommunicationsController'
+// import './UserActionModule'
+// import './PortableExperiences'
+// import './RestrictedActions'
+// import './SignedFetch'
+// import './SceneStateStorageController/SceneStateStorageController'
+// import './Players'
+// import './Permissions'
```
Comment (kuruk-mm) [inline packages/shared/apis/client/index.ts:15]: Can we delete this commented code?

File: packages/shared/apis/host/index.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  registerEthereumControllerServiceServerImplementation(serverPort)
+  registerParcelIdentityServiceServerImplementation(serverPort)
+  registerSocialControllerServiceServerImplementation(serverPort)
+  registerRestrictedActionsServiceServerImplementation(serverPort)
+  registerCommunicationsControllerServiceServerImplementation(serverPort)
+  registerPlayersServiceServerImplementation(serverPort)
+  registerPortableExperiencesServiceServerImplementation(serverPort)
+  registerSignedFetchServiceServerImplementation(serverPort)
+  registerSceneStateStorageControllerServiceServerImplementation(serverPort)
+  // TODO: remove when renderer-rpc is ready
```
Comment (kuruk-mm) [inline packages/shared/apis/host/index.ts:37]: CC https://github.com/decentraland/sdk/issues/321

File: packages/shared/world/SceneWorker.ts [hunk 5]
```diff
@@ -21,17 +15,88 @@ export enum SceneWorkerReadyState {
   DISPOSED = 1 << 8
 }
 
+import { registerServices } from 'shared/apis/host'
+import { PortContext } from 'shared/apis/host/context'
+import { pushableChannel } from '@dcl/rpc/dist/push-channel'
+import Protocol from 'devtools-protocol'
+
+function createGenericLogComponent() {
```
Comment (kuruk-mm) [inline packages/shared/world/SceneWorker.ts:23]: Minor: Some reason this is here in `SceneWorker`? Should not be `createSceneWorkerLogComponent` if only is for the `SceneWorker`?

File: packages/shared/world/parcelSceneManager.ts [hunk 6]
```diff
@@ -4,7 +4,7 @@ import {
   LifecycleManager,
   ParcelSceneLoadingParams
 } from 'decentraland-loader/lifecycle/manager'
-import { ScriptingTransport } from 'decentraland-rpc/lib/common/json-rpc/types'
+// import { ScriptingTransport } from 'decentraland-rpc/lib/common/json-rpc/types'
```
Comment (kuruk-mm) [inline packages/shared/world/parcelSceneManager.ts:7]:
```suggestion
```

File: packages/unity-interface/dcl.ts [hunk 7]
```diff
@@ -90,9 +89,10 @@ async function startGlobalScene(cid: string, title: string, fileContentUrl: stri
     mappings: []
   })
 
-  const worker = loadParcelScene(scene, undefined, true)
+  loadParcelScene(scene, undefined, true)
 
-  await ensureUiApis(worker)
+  // TODO
+  // await ensureUiApis(worker)
```
Comment (kuruk-mm) [inline packages/unity-interface/dcl.ts:95]: `ensureUiApis ` check if `SocialController` host API exists. And looks like you've already created. So maybe this is old?

File: packages/scene-system/sdk/NewSceneRuntime.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+// New
+import { setupStats } from './new-rpc/Stats'
+import { createDecentralandInterface } from './new-rpc/DecentralandInterface'
+import { setupFpsThrottling } from './new-rpc/SetupFpsThrottling'
+import { createEventDispatcher, EventCallback, EventState } from './new-rpc/EventDispatcher'
+import { DevToolsAdapter } from './new-rpc/DevToolsAdapter'
+import type { EntityAction } from 'shared/apis/gen/EngineAPI'
+
+export async function startNewSceneRuntime(client: RpcClient) {
```
Comment (kuruk-mm) [inline packages/scene-system/sdk/NewSceneRuntime.ts:19]: There is a reason to use the word `New`?
The `New` is what it is here. And if we need to maintain old code like `decentraland-rpc`, I could agree to use `old-rpc`.

The same goes for the keyword `new-rpc`. Should be `rpc`, and the old rpc be renamed to `old-rpc`, and its code will be deleted one day.

**Review by kuruk-mm — COMMENTED** [review]
Code reviewed as far as my knowledge about the kernel goes!
I'm testing the branch.

File: packages/scene-system/stateful-scene/RendererStatefulActor.ts [hunk 9]
```diff
@@ -1,15 +1,14 @@
...
+// import {
+//   AttachEntityComponentPayload,
+//   ComponentCreatedPayload,
+//   ComponentRemovedPayload,
+//   ComponentUpdatedPayload,
+//   CreateEntityPayload,
+//   EntityAction,
+//   RemoveEntityPayload,
+//   UpdateEntityComponentPayload
+// } from 'shared/types'
```
Comment (menduz) [inline packages/scene-system/stateful-scene/RendererStatefulActor.ts:11]: Remove commented code
```suggestion
```

File: packages/scene-system/stateful-scene/RendererStatefulActor.ts [hunk 10]
```diff
@@ -20,66 +19,83 @@ import {
...
-    this.eventSubscriber = new EventSubscriber(this.engine)
+  }
+
+  sendBatch(events: EntityAction[]) {
+    this.EngineAPI?.sendBatch({
+      actions: events.map((item) => ({
+        type: item.type,
+        tag: item.tag,
+        payload: item.payload
+      }))
```
Comment (menduz) [inline packages/scene-system/stateful-scene/RendererStatefulActor.ts:44]: why is this map required?

File: packages/shared/apis/IEngineAPI.ts [hunk 11]
```diff
@@ -20,11 +20,6 @@ export interface IEngineAPI {
    */
   sendBatch(actions: EntityAction[]): Promise<void>
 
-  /**
-   * Start signal, sent after everything was initialized
-   */
-  startSignal(): Promise<void>
```
Comment (menduz) [inline packages/shared/apis/IEngineAPI.ts:26]: Where did this go?

File: packages/shared/apis/client/EthereumController.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+     * @internal
+     */
+    async convertMessageToObject(message: string): Promise<MessageDict> {
+      return (await originalService.convertMessageToObject({ message })).dict
+    },
+
+    /**
+     * Used to build a Ethereum provider
+     */
+    async sendAsync(message: RPCSendableMessage): Promise<any> {
```
Comment (menduz) [inline packages/shared/apis/client/EthereumController.ts:60]: Is this used?

File: packages/shared/apis/client/ParcelIdentity.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+          baseUrl: data.land?.baseUrl || '',
+          baseUrlBundles: data.land?.baseUrlBundles || '',
+          mappingsResponse: {
+            root_cid: data.land?.mappingsResponse?.rootCid || '',
+            parcel_id: data.land?.mappingsResponse?.parcelId || '',
+            contents: data.land?.mappingsResponse?.contents || []
+          }
+        },
+        cid: data.cid
+      }
```
Comment (menduz) [inline packages/shared/apis/client/ParcelIdentity.ts:33]: isn't there a function to do exactly this somewhere?

File: packages/shared/apis/client/SignedFetch.ts [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    SignedFetchServiceDefinition
+  )
+
+  return {
+    ...originalService,
+    async signedFetch(url: string, originalInit?: OriginalFlatFetchInit): Promise<OriginalFlatFetchResponse> {
+      let init: FlatFetchInit | undefined = undefined
+      if (originalInit) {
+        init = { headers: {} }
+        if (originalInit.headers && typeof originalInit.headers === 'object') {
```
Comment (menduz) [inline packages/shared/apis/client/SignedFetch.ts:30]: This if doesn't handle the Headers object from `RequestInit`

File: packages/shared/apis/client/index.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (menduz) [inline packages/shared/apis/client/index.ts:15]:
```suggestion
```

File: packages/shared/apis/host/EngineAPI.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+          e.proto ||
+          ({
+            type: EventDataType.Generic,
+            generic: { eventId: e.generic!.type, eventData: JSON.stringify(e.generic!.data) }
+          } as any)
+        )
+      })
+
+      ctx.events = []
+      return { events }
```
Comment (menduz) [inline packages/shared/apis/host/EngineAPI.ts:107]: This may be a bit hard to follow. Could you expand it with named consts?

File: packages/shared/apis/host/Permissions.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+      }
+    }
+
+    if (
+      list.includes(test) &&
+      (test === PermissionItem.ALLOW_TO_MOVE_PLAYER_INSIDE_SCENE ||
+        test === PermissionItem.ALLOW_TO_TRIGGER_AVATAR_EMOTE)
+    ) {
+      return true
+    }
```
Comment (menduz) [inline packages/shared/apis/host/Permissions.ts:50]: Can this conditions be extracted to named consts for clarity?

File: packages/shared/apis/host/PortableExperiences.ts [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+import { PortableExperiencesServiceDefinition } from '../proto/PortableExperiences.gen'
+
+export function registerPortableExperiencesServiceServerImplementation(port: RpcServerPort<PortContext>) {
+  codegen.registerService(port, PortableExperiencesServiceDefinition, async () => ({
+    //   /**
+    //    * Starts a portable experience.
+    //    * @param  {SpawnPortableExperienceParameters} [pid] - Information to identify the PE
+    //    *
+    //    * Returns the handle of the portable experience.
+    //    */
```
Comment (menduz) [inline packages/shared/apis/host/PortableExperiences.ts:22]:
```suggestion
    /**
     * Starts a portable experience.
     * @param  {SpawnPortableExperienceParameters} [pid] - Information to identify the PE
     *
     * Returns the handle of the portable experience.
     */
```

File: packages/shared/apis/host/SignedFetch.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+      const realm = getRealm(state)
+      const isGuest = !!getIsGuestLogin(state)
+      const network = getSelectedNetwork(state)
+
+      const compatibilityRealm:
+        | {
+            domain: string
+            layer: string
+            catalystName: string
+          }
```
Comment (menduz) [inline packages/shared/apis/host/SignedFetch.ts:30]: where does this type comes from?

**Review by kuruk-mm — APPROVED** [review]
Great work! I tested it on Desktop and Web, and it's working smoothly! Congrats!

File: packages/shared/apis/client/EthereumController.ts (hunk 12 above)
Comment (leanmendoza) [reply packages/shared/apis/client/EthereumController.ts:60]: it's exposed in the `decentraland-ecs`

File: packages/shared/apis/IEngineAPI.ts (hunk 11 above)
Comment (leanmendoza) [reply packages/shared/apis/IEngineAPI.ts:26]: it was dead code or related to old `decentraland-rpc`. it was being used nowhere

File: packages/scene-system/stateful-scene/RendererStatefulActor.ts (hunk 10 above)
Comment (leanmendoza) [reply packages/scene-system/stateful-scene/RendererStatefulActor.ts:44]: nop, removing it

File: packages/shared/apis/host/SignedFetch.ts (hunk 19 above)
Comment (leanmendoza) [reply packages/shared/apis/host/SignedFetch.ts:30]: https://github.com/decentraland/kernel/blob/main/packages/shared/apis/SignedFetch.ts#L26

**Review by menduz — APPROVED** [review]

**Review by kuruk-mm — APPROVED** [review]
